File: FinanceiroApp.Core/Services/BankAccountService.cs
using System.Collections.Generic;
using System.Linq;
using FinanceiroApp.Core.Entities;
using FinanceiroApp.Core.ViewModels;

namespace FinanceiroApp.Core.Services
{
  public class BankAccountService : IBankAccountService
  {
    private readonly IDataStore<BankAccount> _bankAccountStore;
    private readonly IDataStore<Bank> _bankStore;
    private readonly ITransactionService _transactionService;

    public BankAccountService(
        IDataStore<BankAccount> bankAccountStore,
        IDataStore<Bank> bankStore,
        ITransactionService transactionService)
    {
      _bankAccountStore = bankAccountStore;
      _bankStore = bankStore;
      _transactionService = transactionService;
    }

    public IEnumerable<BankAccount> GetAllBankAccounts()
    {
      return _bankAccountStore.GetAll()
        .Select(a => new BankAccount
        {
          Id = a.Id,
          Name = a.Name,
          BankId = a.BankId,
          Number = a.Number,
          Agency = a.Agency,
          Balance = a.Balance,
          Bank = _bankStore.GetById(a.BankId)
        });
    }

    public BankAccount GetBankAccountById(int id)
    {
      return _bankAccountStore.GetById(id);
    }

    public BankAccount GetBankAccountByName(string name)
    {
      return _bankAccountStore.GetAll()
          .FirstOrDefault(a => a.Name.Equals(name, StringComparison.OrdinalIgnoreCase));
    }

    public void AddBankAccount(BankAccount bankAccount)
    {
      _bankAccountStore.Add(bankAccount);
    }

    public void UpdateBankAccount(BankAccount bankAccount)
    {
      _bankAccountStore.Update(bankAccount);
    }

    public void DeleteBankAccount(int id)
    {
      _bankAccountStore.Delete(id);
    }

    public IEnumerable<BankAccountViewModel> GetAllBankAccountsWithDetails()
    {
      var bankAccounts = _bankAccountStore.GetAll();
      var banks = _bankStore.GetAll().ToDictionary(b => b.Id);

      return bankAccounts.Select(ba => CreateBankAccountViewModel(ba, banks));
    }

    public decimal GetTotalBalance()
    {
      return GetAllBankAccounts()
          .Sum(account => _transactionService.GetBalanceForAccount(account.Id));
    }

    private BankAccountViewModel CreateBankAccountViewModel(
        BankAccount bankAccount,
        Dictionary<int, Bank> banks)
    {
      var bank = banks.TryGetValue(bankAccount.BankId, out var b) ? b : null;
      var balance = _transactionService.GetBalanceForAccount(bankAccount.Id);

      return new BankAccountViewModel
      {
        Id = bankAccount.Id,
        Name = bankAccount.Name,
        AccountNumber = bankAccount.Number,
        BankId = bankAccount.BankId,
        BankName = bank?.Name ?? "Banco Desconhecido",
        Balance = balance,
        DisplayName = $"{bankAccount.Name} ({bank?.Name ?? "Banco Desconhecido"})"
      };
    }

  }
}

File: FinanceiroApp.FrontEnd/src/app/services/bank-account.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface BankAccount {
  id: number;
  name: string;
  bankId: number;
  bankName: string;
  accountNumber: string;
  balance: number;
  agency: string;
}

@Injectable({
  providedIn: 'root'
})
export class BankAccountService {
  private apiUrl = `${environment.apiUrl}/bankaccounts`;

  constructor(private http: HttpClient) { }

  getAll(): Observable<BankAccount[]> {
    return this.http.get<BankAccount[]>(this.apiUrl);
  }

  getById(id: number): Observable<BankAccount> {
    return this.http.get<BankAccount>(`${this.apiUrl}/${id}`);
  }

  create(bankAccount: Omit<BankAccount, 'id' | 'bankName'>): Observable<BankAccount> {
    return this.http.post<BankAccount>(this.apiUrl, bankAccount);
  }

  update(id: number, bankAccount: Omit<BankAccount, 'id' | 'bankName'>): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/${id}`, bankAccount);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  getBalance(id: number): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/${id}/balance`);
  }
}
